### PR TITLE
[Merged by Bors] - Feat: port algebra.regular.smul

### PR DESCRIPTION
**Pull request by arienmalec**
Port of algebra.regular.smul

Based on 550b58538991c8977703fdeb7c9d51a5aa27df11

### CODE REVIEW

**Comment by kbuzzard** [comment]
Should this file be called `Smul.lean` or `SMul.lean`? We have two contrasting conventions in this PR.

**Comment by arienmalec** [comment]
> Should this file be called Smul.lean or SMul.lean? We have two contrasting conventions in this PR.

Should be `SMul.lean`; I created it wrongly, changed it locally, but the rename didn't take, so I changed it on Github and hopefully it took?

File: Mathlib/Algebra/Regular/SMul.lean [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+is a commutative ring `R` acting an a module `M`. Since the properties are "multiplicative", there
+is no actual requirement of having an addition, but there is a zero in both `R` and `M`.
+Smultiplications involving `0` are, of course, all trivial.
+
+The defining property is that an element `a ∈ R` is `M`-regular if the smultiplication map
+`M → M`, defined by `m ↦ a • m`, is injective.
+
+This property is the direct generalization to modules of the property `IsLeftRegular` defined in
+`Algebra/Regular`.  Lemma `IsSMulRegular.is_left_regular_iff` shows that indeed the two notions
+coincide.
```
Comment (dupuisf) [inline Mathlib/Algebra/Regular/SMul.lean:30]:
```suggestion
`Algebra/Regular`.  Lemma `isLeftRegular_iff` shows that indeed the two notions
coincide.
```

File: Mathlib/Algebra/Regular/SMul.lean [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+end HasSmul
+
+section Monoid
+
+variable [Monoid R] [MulAction R M]
+
+variable (M)
+
+/-- One is `M`-regular always. -/
```
Comment (dupuisf) [inline Mathlib/Algebra/Regular/SMul.lean:136]:
```suggestion
/-- One is always `M`-regular. -/
```

File: Mathlib/Algebra/Regular/SMul.lean [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+end CommSemigroup
+
+end IsSMulRegular
+
+section Group
+
+variable {G : Type _} [Group G]
+
+/-- An element of a group acting on a Type is regular. This relies on the availability
+of the inverse given by groups, since there is no `left_cancel_smul` typeclass. -/
```
Comment (dupuisf) [inline Mathlib/Algebra/Regular/SMul.lean:238]:
```suggestion
of the inverse given by groups, since there is no `LeftCancelSMul` typeclass. -/
```

**Review by dupuisf — COMMENTED** [review]
Just a few docstring changes, otherwise LGTM!

bors d+

**Pull request closed without merging**